File: cli/src/main/kotlin/com/maestrorecorder/cli/recorder/RecordingSession.kt
package com.maestrorecorder.cli.recorder

import com.maestrorecorder.cli.adb.AdbBridge
import com.maestrorecorder.cli.adb.DeviceManager
import com.maestrorecorder.cli.adb.InputDeviceDiscovery
import com.maestrorecorder.cli.generator.YamlGenerator
import com.maestrorecorder.cli.grpc.AgentClient
import com.maestrorecorder.cli.merger.EventMerger
import com.maestrorecorder.cli.parser.CoordinateConverter
import com.maestrorecorder.cli.parser.GeteventParser
import com.maestrorecorder.shared.models.MaestroCommand
import com.maestrorecorder.shared.models.TapOnSelector
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.delay
import java.util.concurrent.atomic.AtomicBoolean
import kotlin.io.path.writeText

class RecordingSession(private val config: SessionConfig) {

    private val adb = AdbBridge()
    private val deviceManager = DeviceManager(adb)
    private val running = AtomicBoolean(true)

    suspend fun start() {
        // 1. Device discovery
        val deviceSerial = config.deviceSerial ?: deviceManager.getConnectedDevice()
        printStatus("Connected to device: $deviceSerial")

        // 2. Get screen size
        val (screenWidth, screenHeight) = deviceManager.getScreenSize()
        printStatus("Screen size: ${screenWidth}x${screenHeight}")

        // 3. Discover touchscreen input device and ranges
        val inputDevice = InputDeviceDiscovery.discover(adb)
        printStatus("Touchscreen: ${inputDevice.devicePath} (max: ${inputDevice.maxX}x${inputDevice.maxY})")

        // 4. Set up port forwarding
        deviceManager.forwardPort(config.grpcPort, config.grpcPort)
        printStatus("Forwarding port: adb forward tcp:${config.grpcPort} tcp:${config.grpcPort}")

        // 5. Wait for GRPC agent to be ready
        val client = waitForAgent()
        printStatus("GRPC agent connected")

        // 6. Determine app ID
        val appId = config.appId ?: detectForegroundApp()
        printStatus("Target app: $appId")

        // 7. Create coordinate converter
        val converter = CoordinateConverter(
            inputMaxX = inputDevice.maxX,
            inputMaxY = inputDevice.maxY,
            screenWidth = screenWidth,
            screenHeight = screenHeight
        )

        // 8. Start getevent stream
        val (geteventFlow, geteventProcess) = adb.stream("shell", "getevent", "-lt", inputDevice.devicePath)

        // 9. Set up shutdown hook
        Runtime.getRuntime().addShutdownHook(Thread {
            running.set(false)
            geteventProcess.destroy()
        })

        // 10. Set up pipeline
        val parser = GeteventParser(converter, deviceFilter = inputDevice.devicePath)
        val merger = EventMerger(client)
        val commands = mutableListOf<MaestroCommand>(MaestroCommand.LaunchApp)

        printStatus("Recording... (press Ctrl+C to stop)\n")

        // 11. Collect merged commands
        try {
            merger.merge(parser.parse(geteventFlow)).collect { command ->
                commands.add(command)
                printAction(command)
            }
        } catch (_: CancellationException) {
            // Normal shutdown
        } catch (_: Exception) {
            // Stream ended (process destroyed)
        }

        // 12. Generate YAML
        val generator = YamlGenerator()
        val yaml = generator.generate(appId, commands)
        config.outputFile.writeText(yaml)

        // 13. Cleanup
        client.close()
        geteventProcess.destroy()
        deviceManager.removePortForward(config.grpcPort)

        printStatus("\nRecording stopped. ${commands.size} actions captured.")
        printStatus("Generated: ${config.outputFile}")
        printStatus("Run with: maestro test ${config.outputFile}")
    }

    private suspend fun waitForAgent(): AgentClient {
        repeat(10) { attempt ->
            try {
                val client = AgentClient(port = config.grpcPort)
                client.getDeviceInfo() // probe call
                return client
            } catch (_: Exception) {
                if (attempt < 9) {
                    delay(500)
                }
            }
        }
        throw IllegalStateException("Agent GRPC server did not start within 5 seconds. Make sure the agent is running.")
    }

    private suspend fun detectForegroundApp(): String {
        return try {
            val output = adb.exec("shell", "dumpsys", "activity", "activities")
            val regex = Regex("""mResumedActivity.*?(\S+)/""")
            regex.find(output)?.groupValues?.get(1) ?: "com.unknown.app"
        } catch (_: Exception) {
            "com.unknown.app"
        }
    }

    private fun printAction(command: MaestroCommand) {
        val desc = when (command) {
            is MaestroCommand.LaunchApp -> "LAUNCH"
            is MaestroCommand.TapOn -> when (val s = command.selector) {
                is TapOnSelector.ById -> "TAP → id:${s.id}"
                is TapOnSelector.ByText -> "TAP → \"${s.text}\""
                is TapOnSelector.ByContentDescription -> "TAP → [${s.description}]"
                is TapOnSelector.ByPoint -> "TAP → (${s.x}, ${s.y})"
            }
            is MaestroCommand.InputText -> "INPUT → \"${command.text}\""
        }
        System.err.println("  $desc")
    }

    private fun printStatus(message: String) {
        System.err.println("▶ $message")
    }
}

File: packages/server/src/recording/recording-session.ts
import { EventEmitter } from 'events';
import { execFile } from 'child_process';
import type { UserAction, MaestroCommand, DeviceInfo } from '@maestro-recorder/shared';
import { GeteventStream, discoverInputDevice } from './getevent-parser.js';
import { TouchStateMachine } from './touch-state-machine.js';
import { CoordinateConverter } from './coordinate-converter.js';
import { AgentClient } from './agent-client.js';
import { EventMerger } from './event-merger.js';
import { YamlGenerator } from './yaml-generator.js';
import type { GeteventLine } from './types.js';

export class RecordingSession extends EventEmitter {
  private geteventStream: GeteventStream | null = null;
  private touchStateMachine = new TouchStateMachine();
  private converter: CoordinateConverter | null = null;
  private agent: AgentClient;
  private merger: EventMerger | null = null;
  private yamlGenerator = new YamlGenerator();
  private _commands: MaestroCommand[] = [];
  private deviceInfo: DeviceInfo | null = null;

  get commands(): MaestroCommand[] { return this._commands; }

  constructor(
    private deviceSerial: string,
    private appId: string,
    private agentPort = 50051,
  ) {
    super();
    this.agent = new AgentClient(agentPort);
  }

  async start(): Promise<void> {
    // 1. Setup port forwarding
    await this.adbExec('forward', `tcp:${this.agentPort}`, `tcp:${this.agentPort}`);

    // 2. Check agent
    const agentAlive = await this.agent.ping();
    if (!agentAlive) {
      throw new Error(
        'Agent not running on device. Start it with:\n' +
        'adb shell am instrument -w -e class com.maestrorecorder.agent.RecorderInstrumentation#startServer ' +
        'com.maestrorecorder.agent.test/androidx.test.runner.AndroidJUnitRunner'
      );
    }

    // 3. Get device info
    this.deviceInfo = await this.agent.deviceInfo();
    if (!this.deviceInfo) {
      // Fallback to ADB
      const sizeOutput = await this.adbExec('shell', 'wm', 'size');
      const m = sizeOutput.match(/(\d+)x(\d+)/);
      this.deviceInfo = {
        screenWidth: m ? parseInt(m[1]) : 1080,
        screenHeight: m ? parseInt(m[2]) : 1920,
        density: 420,
      };
    }

    // 4. Discover input device
    const inputDevice = await discoverInputDevice(this.deviceSerial);

    // 5. Setup coordinate converter
    this.converter = new CoordinateConverter(
      inputDevice.maxX, inputDevice.maxY,
      this.deviceInfo.screenWidth, this.deviceInfo.screenHeight,
    );

    // 6. Setup event merger
    this.merger = new EventMerger(this.agent, this.deviceInfo.screenWidth, this.deviceInfo.screenHeight);

    // 7. Add launchApp as first command
    this.addCommand({ type: 'launchApp' });

    // 8. Start getevent stream
    this.geteventStream = new GeteventStream(this.deviceSerial, inputDevice.devicePath);

    this.geteventStream.on('line', (line: GeteventLine) => {
      // Emit raw getevent line
      this.emit('raw', line);

      const action = this.touchStateMachine.feed(line, this.converter!);
      if (action) {
        this.onUserAction(action);
      }
    });

    this.geteventStream.on('error', (err: Error) => {
      this.emit('error', err.message);
    });

    this.geteventStream.start();
    this.emit('status', 'recording');
  }

  private async onUserAction(action: UserAction): Promise<void> {
    if (!this.merger) return;

    // Emit raw parsed action (for Parsed tab)
    this.emit('action', action);

    try {
      const { command, element } = await this.merger.merge(action);

      // Emit element data (for Element tab)
      if (element) {
        this.emit('element', { action, element });
      }

      this.addCommand(command);

      // Detect text input for taps on text fields
      if (action.type === 'tap' && element) {
        const text = await this.merger.detectTextInput(element, action.x, action.y);
        if (text) {
          this.addCommand({ type: 'inputText', text });
        }
      }
    } catch (err) {
      // Skip failed merges silently
    }
  }

  private addCommand(command: MaestroCommand): void {
    this._commands.push(command);
    this.emit('command', this.commandToDto(command));
  }

  async stop(): Promise<{ yaml: string; commands: MaestroCommand[] }> {
    this.geteventStream?.stop();
    this.geteventStream = null;

    // Remove port forward
    try {
      await this.adbExec('forward', '--remove', `tcp:${this.agentPort}`);
    } catch {}

    const yaml = this.yamlGenerator.generate(this.appId, this._commands);

    this.emit('status', 'stopped');
    return { yaml, commands: this._commands };
  }

  private commandToDto(cmd: MaestroCommand): any {
    if (cmd.type === 'tapOn' || cmd.type === 'longPressOn' || cmd.type === 'doubleTapOn' ||
        cmd.type === 'assertVisible' || cmd.type === 'assertNotVisible' || cmd.type === 'scrollUntilVisible') {
      const sel = (cmd as any).selector;
      return {
        type: cmd.type,
        selector: {
          type: sel.kind === 'id' ? 'ById' : sel.kind === 'text' ? 'ByText' :
                sel.kind === 'contentDescription' ? 'ByContentDescription' : 'ByPoint',
          value: sel.kind === 'point' ? `${sel.x},${sel.y}` :
                 sel.kind === 'id' ? sel.id : sel.kind === 'text' ? sel.text : sel.description,
        },
      };
    }
    if (cmd.type === 'inputText') return { type: 'InputText', text: cmd.text };
    if (cmd.type === 'swipe') return { type: 'Swipe', ...cmd };
    return { type: cmd.type };
  }

  private adbExec(...args: string[]): Promise<string> {
    return new Promise((resolve, reject) => {
      execFile('adb', ['-s', this.deviceSerial, ...args], (err, stdout) => {
        if (err) reject(err);
        else resolve(stdout.trim());
      });
    });
  }
}
